viadot/sources/supermetrics.py

```diff
@@ -0,0 +1,174 @@
+        # Supermetrics allows pivoting GA data, in which case it generates additional columns,
+        # which are not enlisted in response's query metadata but are instead added as the first row of data.
+        is_pivoted = any(
+            field["field_split"] == "column"
+            for field in response["meta"]["query"]["fields"]
+        )
+
+        if is_pivoted:
+            if not response["data"]:
+                raise ValueError(
+                    "Couldn't find column names as query returned no data."
+                )
+            columns = response["data"][0]
+        else:
+            # non-pivoted data; query fields match result fields
+            cols_meta = response["meta"]["query"]["fields"]
+            columns = [col_meta["field_name"] for col_meta in cols_meta]
+        return columns
+
+    @classmethod
+    def _get_col_names_other(cls, response: dict) -> List[str]:
+        """
+        Returns list of columns names (to Google Analytics use  _get_col_names_google_analytics().
+
+        Args:
+            response (dict):  Dictionary with the json response from API call.
+
+        """
+        cols_meta = response["meta"]["query"]["fields"]
+        columns = [col_meta["field_name"] for col_meta in cols_meta]
+        return columns
+
+    def _get_col_names(self) -> List[str]:
+        """
+        Returns list of columns names.
+
+        Args:
+            None.
+
+        """
+        query_params_cp = deepcopy(self.query_params)
+        query_params_cp["offset_start"] = 0
+        query_params_cp["offset_end"] = 0
+
+        response: dict = Supermetrics(query_params=query_params_cp).to_json()
+        if self.query_params["ds_id"] == "GA":
+            return Supermetrics._get_col_names_google_analytics(response)
+        else:
+            return Supermetrics._get_col_names_other(response)
+
+    def to_df(self, if_empty: str = "warn") -> pd.DataFrame:
+        """
+        Download data into a pandas DataFrame.
+
+        Note that Supermetric can calculate some fields on the fly and alias them in the
+        returned result. For example, if the query requests the `position` field,
+        Supermetric may return an `Average position` caclulated field.
+        For this reason we take columns names from the actual results rather than from input fields.
+
+        Args:
+            if_empty (str, optional): What to do if query returned no data. Defaults to "warn".
+
+
+        """
+        try:
+            columns = self._get_col_names()
+        except ValueError:
+            columns = None
+
+        data = self.to_json()["data"]
+
+        if data:
+            df = pd.DataFrame(data[1:], columns=columns).replace("", np.nan)
+        else:
+            df = pd.DataFrame(columns=columns)
+
+        if df.empty:
+            self._handle_if_empty(if_empty)
+
+        return df
+
+    def query(self, params: Dict[str, Any]):
+        """
+        Returns the query with the credentials info.
```

Comment:
The name of the function/method should describe what it does, for example, "Query the database using parameters." Then it should describe what parameters can be set and which of them are the default parameters. As example - `api_key` parameter is missed in the description.

Returned values should be described on bottom of docsting, like with `Args:`, but `Returns:`